mirror_builder/sources.py

```diff
@@ -0,0 +1,43 @@
+import logging
+
+import resolvelib
+from packaging.requirements import Requirement
+
+from . import overrides, resolve_and_download
+
+logger = logging.getLogger(__name__)
+
+
+def download_source(ctx, req):
+    logger.info('downloading source for %s', req)
+    r = Requirement(req)
+    downloader = overrides.find_override_method(r.name, 'download_source')
+    if not downloader:
+        downloader = _default_download_source
+    source_filename = downloader(ctx, req)
+    logger.info('downloaded source for %s to %s', req, source_filename)
+    return source_filename
+
+
+def _default_download_source(ctx, req):
+    "Download the requirement and return the name of the output path."
+    r = Requirement(req)
+
+    # Create the (reusable) resolver.
+    provider = resolve_and_download.PyPIProvider()
+    reporter = resolve_and_download.BaseReporter()
+    resolver = resolvelib.Resolver(provider, reporter)
+
+    # Kick off the resolution process, and get the final result.
+    logger.debug("resolving requirement %s", req)
+    try:
+        result = resolver.resolve([r])
+    except (resolvelib.InconsistentCandidate,
+            resolvelib.RequirementsConflicted,
+            resolvelib.ResolutionImpossible) as err:
+        logger.warning(f'could not resolve {req}: {err}')
+    else:
+        return resolve_and_download.download_resolution(
```

Comment:
Even for something like PyTorch I can see us wanting to resolve a specific version based on a requirements specification that includes markers and extras. We'll need to split the "resolve" part out of "resolve_and_download" to make that possible. We should do that in a separate PR for clarity.

Reply:
Yeah, and we need to be able to resolve without sdists - see `PyPIProvider(only_sdists=False)`